validation_to_html.py

```diff
@@ -143,104 +136,128 @@ def esc(s):
+        continents[c.continent] = c + continents.get(c.continent, CityData())
+        if c.continent not in c_by_c:
+            c_by_c[c.continent] = set()
+        c_by_c[c.continent].add(c.country)
+    world = sum(continents.values(), CityData())
+
+    overground = "traml_expected" in next(iter(data.values())).data
+    date = datetime.datetime.utcnow().strftime("%d.%m.%Y %H:%M UTC")
+    index = open(os.path.join(target_dir, "index.html"), "w", encoding="utf-8")
+    index.write(tmpl(INDEX_HEADER, world))
+
+    for continent in sorted(continents.keys()):
+        content = ""
+        for country in sorted(c_by_c[continent]):
+            country_file_name = country.lower().replace(" ", "-") + ".html"
+            content += tmpl(
+                INDEX_COUNTRY,
+                countries[country],
+                file=country_file_name,
                 country=country,
                 continent=continent,
-                overground=overground,
-                subways=not overground,
             )
-        )
-        for name, city in sorted(data.items()):
-            if city.country == country:
-                file_base = os.path.join(path, city.slug)
-                yaml_file = (
-                    city.slug + ".yaml"
-                    if os.path.exists(file_base + ".yaml")
-                    else None
-                )
-                json_file = (
-                    city.slug + ".geojson"
-                    if os.path.exists(file_base + ".geojson")
-                    else None
-                )
-                errors = "<br>".join([osm_links(esc(e)) for e in city.errors])
-                warnings = "<br>".join(
-                    [osm_links(esc(w)) for w in city.warnings]
-                )
-                notices = "<br>".join(
-                    [osm_links(esc(n)) for n in city.notices]
+            country_file = open(
+                os.path.join(target_dir, country_file_name),
+                "w",
+                encoding="utf-8",
+            )
+            country_file.write(
+                tmpl(
+                    COUNTRY_HEADER,
+                    country=country,
+                    continent=continent,
+                    overground=overground,
+                    subways=not overground,
                 )
-                country_file.write(
-                    tmpl(
-                        COUNTRY_CITY,
-                        city,
-                        city=name,
-                        country=country,
-                        continent=continent,
-                        yaml=yaml_file,
-                        json=json_file,
-                        subways=not overground,
-                        errors=errors,
-                        warnings=warnings,
-                        notices=notices,
-                        overground=overground,
+            )
+            for name, city in sorted(data.items()):
+                if city.country == country:
```

Comment:
You can reverse the condition, using continue

Reply:
\+
The filter by country has been moved to the earlier stage (to the generator expression), resulting in sorting smaller number of items. PTAL.